webrender/src/prim_store.rs

```diff
@@ -1200,44 +1166,57 @@ impl PrimitiveStore {
         let mut prim_needs_resolve = false;
         let mut rebuild_bounding_rect = false;
 
-        if let GpuLocation::GpuCache(ref mut cache_id) = metadata.gpu_location {
-            let cpu_borders = &self.cpu_borders;
-            let cpu_box_shadows = &self.cpu_box_shadows;
-            let cpu_images = &self.cpu_images;
-            let cpu_yuv_images = &self.cpu_yuv_images;
-            let cpu_rectangles = &self.cpu_rectangles;
-            let cpu_prim_index = metadata.cpu_prim_index;
-            let prim_kind = metadata.prim_kind;
-
-            // Mark this GPU resource as required for this frame.
-            resource_cache.gpu_cache.request(cache_id, |blocks| {
-                match prim_kind {
-                    PrimitiveKind::Rectangle => {
-                        let rect = &cpu_rectangles[cpu_prim_index.0];
-                        rect.write_gpu_blocks(blocks);
-                    }
-                    PrimitiveKind::Border => {
-                        let border = &cpu_borders[cpu_prim_index.0];
-                        border.write_gpu_blocks(blocks);
-                    }
-                    PrimitiveKind::BoxShadow => {
-                        let box_shadow = &cpu_box_shadows[cpu_prim_index.0];
-                        box_shadow.write_gpu_blocks(blocks);
-                    }
-                    PrimitiveKind::Image => {
-                        let image = &cpu_images[cpu_prim_index.0];
-                        image.write_gpu_blocks(blocks);
-                    }
-                    PrimitiveKind::YuvImage => {
-                        let yuv_image = &cpu_yuv_images[cpu_prim_index.0];
-                        yuv_image.write_gpu_blocks(blocks);
-                    }
-                    _ => {
-                        unreachable!("Encountered a type not supported by GPU cache!");
-                    }
+        // TODO(gw): Restructure this code so the borrow checker
+        //           doesn't force the locals below!
```

Comment:
oh my, this is getting out of hand!
One option, perhaps the most idiomatic one, would be to have a custom iterator returned by `request_iter()`, so that we can iterate it without a closure